rclcpp/src/rclcpp/executor.cpp

```diff
@@ -275,7 +275,7 @@ Executor::spin_until_future_complete_impl(
   if (spinning.exchange(true)) {
     throw std::runtime_error("spin_until_future_complete() called while already spinning");
   }
-  RCPPUTILS_SCOPE_EXIT(this->spinning.store(false); );
+  RCPPUTILS_SCOPE_EXIT(this->spinning.store(false);wait_result_.reset(););
```

Comment:
this order is not thread-safe.
you are first setting `spinning` to false and then resetting the wait_result.

If there are two threads, one trying to start spin and the other exiting spin, there can be a concurrent access to `wait_result_`.
We should at least invert the order, to ensure that no other threads can enter spin until we have cleared the wait result.

Reply:
Thanks. It is better to invert the order.  

Besides, There will be many issues if multiple threads executes the same executor's spin concurrently.  
`wait_result_` was completely unprotected.

https://github.com/ros2/rclcpp/blob/f4923c6f43afaf983f95158584d37687f2679925/rclcpp/src/rclcpp/executor.cpp#L724-L751

Reply:
At least in the case of the multithreaded executor, the wait_result is protected by a mutex in the mt executor itself (https://github.com/ros2/rclcpp/blob/f4923c6f43afaf983f95158584d37687f2679925/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L88)

As for single-threaded executor, it should generally not be possible to call spin from multiple threads as there is an internal `spinning` atomic that should block other threads from spinning an executor that is already spinning.

Reply:
@mjcarroll Thanks for your comments.  
I didn't express myself clearly. I mean, For the SingleThreadedExecutor, if the user creates multiple threads to execute spin(), problems will arise.